platforms/allwinner-d1/boards/lichee-rv/src/uart.rs

```diff
@@ -0,0 +1,181 @@
+use core::{
+    ptr::{null_mut, NonNull},
+    sync::atomic::{AtomicPtr, Ordering},
+};
+
+use d1_pac::UART0;
+use drivers::dmac::{
+    descriptor::{
+        AddressMode, BModeSel, BlockSize, DataWidth, DescriptorConfig, DestDrqType, SrcDrqType,
+    },
+    Channel, ChannelMode,
+};
+use kernel::{
+    comms::{
+        bbq::{new_bidi_channel, BidiHandle, Consumer, GrantW, SpscProducer},
+        kchannel::{KChannel, KConsumer},
+    },
+    maitake::sync::WaitCell,
+    registry::{
+        simple_serial::{Request, Response, SimpleSerial, SimpleSerialError},
+        Message,
+    },
+    Kernel,
+};
+
+struct GrantWriter {
+    grant: GrantW,
+    used: usize,
+}
+
+impl core::fmt::Write for GrantWriter {
+    fn write_str(&mut self, s: &str) -> core::fmt::Result {
+        let glen = self.grant.len();
+        let slen = s.len();
+        let new_used = self.used + slen;
+        if new_used <= glen {
+            self.grant[self.used..][..slen].copy_from_slice(s.as_bytes());
+            self.used = new_used;
+            Ok(())
+        } else {
+            Err(core::fmt::Error)
+        }
+    }
+}
+
+static TX_DONE: WaitCell = WaitCell::new();
+static UART_RX: AtomicPtr<SpscProducer> = AtomicPtr::new(null_mut());
+
+pub struct D1Uart {
+    _x: (),
+}
+
+impl D1Uart {
+    pub fn tx_done_waker() -> &'static WaitCell {
+        &TX_DONE
+    }
+
+    pub fn handle_uart0_int() {
+        let uart0 = unsafe { &*UART0::PTR };
+        let prod = UART_RX.load(Ordering::Acquire);
+        if !prod.is_null() {
+            let prod = unsafe { &*prod };
+
+            while let Some(mut wgr) = prod.send_grant_max_sync(64) {
+                let used_res = wgr.iter_mut().enumerate().try_for_each(|(i, b)| {
+                    if uart0.usr.read().rfne().bit_is_set() {
+                        *b = uart0.rbr().read().rbr().bits();
+                        Ok(())
+                    } else {
+                        Err(i)
+                    }
+                });
+
+                match used_res {
+                    Ok(()) => {
+                        let len = wgr.len();
+                        wgr.commit(len);
+                    }
+                    Err(used) => {
+                        wgr.commit(used);
+                        break;
+                    }
+                }
+            }
```

Comment:
nit, TIOLI: this logic kinda feels like it could maybe be simplified with a for loop:
```suggestion
            'read: while let Some(mut wgr) = prod.send_grant_max_sync(64) {
                for (used, b) in wgr.iter_mut().enumerate() {
                    if !uart0.usr.read().rfne().bit_is_set() {
                        wgr.commit(used);
                        break 'read;
                    }

                    *b = uart0.rbr().read().rbr().bits();
                }

                let len = wgr.len();
                wgr.commit(len);
            }
```
which, AFAICT, should do the same thing?